### ISSUE TITLE
Support for edge runtimes like cloudflare workers, vercel edge, deno etc

### ISSUE THREAD

**Issue by cosbgn** [readmeio/oas#825]
As a feature request, it would be amazing if this library could work on `cloudflare workers`,  `deno`, `vercel-edge`, `netlify-edge` and other edge runtimes. 

These runtimes are gaining a lot of popularity lately, but unfortunately can't be used with this library. I guess this is also due to dependancies like `ajv`, but there are some edge ready alternatives like `@cfworker/json-schema`.

Thanks for considering it.

**Comment by erunion** [comment]
I would love to support Edge runtimes! Do you have a solid way of detecting a package that doesn't support it?

**Comment by cosbgn** [comment]
Not really, I think the easiest is just to use `wrangler dev` locally and see if an error comes up.

So far I've learned that: 

- Imports needs to be prefixed (e.g. `import { Buffer } from 'node:buffer'`)
- The `nodejs_compat` flag helps a lot, can easily be activated on cloudflare and active by default on vercel-edge
- ajv doesn't work, but  @cfworker/json-schema does
- `eval(), new Function(), setTimeout([string]), and setInterval([string])` can't be used.

**Comment by erunion** [comment]
Do you have an example code snippet for `oas` that's not working with `wrangler dev`? Would be helpful to see and debug. I know we only use `ajv` in `oas/reducer`, which isn't exported from any main export, and only *really* exists in this repository because it's used in https://github.com/readmeio/rdme

**Comment by cosbgn** [comment]
This is the started wrangler which fails on `npm run start`: https://github.com/cosbgn/oas-edge

**Comment by erunion** [comment]
Ah gotcha, so its failing on `oas-normalize` likely then. Moving that off `ajv` is going to be a lot of work because we'll have to move [@readme/openapi-parser](https://npm.im/@readme/openapi-parser) and then rewrite [@readme/better-ajv-errors](https://npm.im/@readme/better-ajv-errors) as well. 😓

**Comment by cosbgn** [comment]
`oas-normalize` would be great as well, but it's failing also without it, only with `oas` check the new commit without normalize: https://github.com/cosbgn/oas-edge/blob/main/src/index.js

**Comment by erunion** [comment]
Woof, ok I see now.

![Screen Shot 2023-10-24 at 1 27 22 PM](https://github.com/readmeio/oas/assets/33762/675381b4-d995-48d6-9024-e31565025c4f)

**Comment by kanadgupta** [comment]
@cosbgn I know it's been a while, but a few things have changed so I think we might be in good shape now:

- We recently overhauled our parsing packages so I believe all of our URL resolvers in `json-schema-ref-parser` are backed by `fetch`
- All of the runtimes in question (i.e., [Cloudflare](https://developers.cloudflare.com/workers/runtime-apis/nodejs/), [Vercel](https://vercel.com/docs/functions/runtimes/edge#compatible-node.js-modules), and [Deno](https://docs.deno.com/api/node/), the latter of which powers Netlify's edge runtime) have pretty solid compatibility with most (if not all?) of the Node APIs we need

Are you still blocked on this?

**Comment by cosbgn** [comment]
Hey @kanadgupta I moved to a different setup because it was incompatible so I'm not using or blocked by this. 

This said it's great if you get it to work on the edge, I might revisit it again soon.

**Comment by erunion** [comment]
Been a hot second since an update on this but with the recent rewrite or our OpenAPI parser that I did earlier this year we are now fully supported within edge runtimes!

Re-running your https://github.com/cosbgn/oas-edge example repository with the latest releases of `oas` and `oas-normalize`:

<img width="1301" alt="Image" src="https://github.com/user-attachments/assets/7445bcc8-59ed-4574-8081-6523e475a40a" />

And out of curiosity I also created a new project with Wrangler v4 and we work under that too!

<img width="1301" alt="Image" src="https://github.com/user-attachments/assets/14d2224c-3be0-43db-a788-92a2cd8d6769" />
